### ISSUE TITLE
Potential issue with activity mocking and cancellation

### ISSUE THREAD

**Issue by cretz** [temporalio/sdk-go#778]
A couple of issues have potentially been observed with the workflow/activity test env:

1. A mocked activity may not return a value if the context is cancelled
2. An activity called on the cancelled context for cleanup is immediately returning and still running the activity

Need standalone replication, then debugging/fix if applicable.

**Comment by awsaba** [comment]
> 2. An activity called on the cancelled context for cleanup is immediately returning and still running the activity

Our use case is not so much clean up as multiple steps of bookeeping. Putting a `switch` after every return to distinguish between an `ActivityError` and cancellation seems like the obvious but wrong solution since it would have to added after most activities.

The workflows and test:
```go
package cancelwork

import (
	"context"
	"fmt"
	"time"

	"go.temporal.io/sdk/activity"
	"go.temporal.io/sdk/workflow"
)

// CancelMe is long-running activity that is the most likely to be running
// when a cancel is received
func CancelMe(ctx context.Context, _ string) (string, error) {
	time.Sleep(1 * time.Second)
	return "long-running-result", nil
}

// SaveStatus just logs the status, but could do some database operation
func SaveStatus(ctx context.Context, value string) error {
	logger := activity.GetLogger(ctx)
	time.Sleep(3 * time.Second)
	logger.Error(fmt.Sprintf("from SaveStatus, value is: \"%s\"", value))
	return nil
}

// SomeWorkflow calls a long running op to do something
func SomeWorkflow(ctx workflow.Context) error {
	logger := workflow.GetLogger(ctx)
	ao := workflow.ActivityOptions{
		StartToCloseTimeout:    10 * time.Minute,
		ScheduleToCloseTimeout: 24 * time.Hour,
		WaitForCancellation:    true,
	}
	ctx = workflow.WithActivityOptions(ctx, ao)

	defer func() {
		logger.Error("some other log before return")
	}()

	var val string
	err := workflow.ExecuteActivity(ctx, CancelMe, "").Get(ctx, &val)
	if err != nil {
		logger.Error(fmt.Sprintf("error in CancelMe: %s, is there a way to get val: \"%s\"", err.Error(), val))
		perr := workflow.ExecuteActivity(ctx, SaveStatus, val).Get(ctx, nil)
		if perr != nil {
			logger.Error(fmt.Sprintf("if cancelled, this should fail with cancel and not execute SaveStatus: \"%s\"", perr.Error()))
		}
		return err
	}

	return nil
}
```

```
import (
	"context"
	"testing"
	"time"

	"github.com/stretchr/testify/mock"
	"github.com/stretchr/testify/suite"
	"go.temporal.io/sdk/testsuite"
)

type CancelMeSuite struct {
	suite.Suite
	testsuite.WorkflowTestSuite
}

func TestCancel(t *testing.T) {
	suite.Run(t, &CancelMeSuite{})
}

func (s *CancelMeSuite) TestCancel() {
	env := s.NewTestWorkflowEnvironment()
	env.OnActivity(CancelMe, mock.AnythingOfType("*context.timerCtx"), "").
		Return(func(ctx context.Context, _ string) (string, error) {
			env.CancelWorkflow()
			return "mocked-return", nil
		}).After(time.Minute)

	env.RegisterWorkflow(SomeWorkflow)
	env.RegisterActivity(CancelMe)
	env.RegisterActivity(SaveStatus)

	env.ExecuteWorkflow(SomeWorkflow)
	s.True(env.IsWorkflowCompleted())
	s.Error(env.GetWorkflowError())

	env.AssertExpectations(s.T())
        // sleep to see the output from SaveStatus
	time.Sleep(5 * time.Second)
}
```
 
The log output will log the error, and then eventually print out the log message from the 2nd activity (`SaveStatus`) called with the already cancelled context.

**Comment by awsaba** [comment]
Follow up from Slack thread:

> A couple of issues have potentially been observed with the workflow/activity test env:
> 
>     1. A mocked activity may not return a value if the context is cancelled
> 
Based on Maxim's responses in the Slack thread, this is expected. For our use case that would need the returned `&val` for a cleanup activity using a disconnected context, we are going to think about how else we can structure that.

>     2. An activity called on the cancelled context for cleanup is immediately returning and still running the activity
> 
>> The call is going to fail immediately without ever generating a ScheduleActivityTask command.

I think this means that `ExecuteActivity` in a workflow test should _not_ be executing the activity, WDYT?

**Comment by cretz** [comment]
To clarify here... Based on my tests, if you don't respect cancellation in the activity, you won't return from `ExecuteActivity` here. You must heartbeat in activities to receive cancellation and you'll probably want a `HeartbeatTimeout` so you don't have to wait 15 seconds.

So basically if your activity was:

```go
func CancelMe(ctx context.Context, _ string) (string, error) {
	for {
		activity.RecordHeartbeat(ctx)
		select {
		case <-ctx.Done():
			return "long-running-result", nil
		case <-time.After(1 * time.Second):
		}
	}
}
```

Then `ExecuteActivity` would not return an error on cancel, it'd return the expected value. If you didn't have that mechanism but instead ignored the concept of heartbeat and cancellation, your activity could run until the end of time and cancellation would have no affect on it.

> I think this means that `ExecuteActivity` with a cancelled context in a workflow test should _not_ be executing the activity, WDYT?

Agreed. I will open a PR for this.

**Comment by cretz** [comment]
I have opened #780 to make sure activities invoked with a cancelled context are not started.

**Comment by cretz** [comment]
We have merged #780 which has some fixes for this area. The next release may be in ~2 weeks. I am closing now, but reopen or comment if you think the PR doesn't address any specific activity + cancellation + test suite issues.

**Comment by awsaba** [comment]
@cretz Thanks!
